Update coredistools

BruceForstall opened the pull request:
Update to the 1.4.0 version of the coredistools package.

Also, implement a basic late disassembler for RyuJIT based on
coredistools (previously, this depended on closed-source msvcdis).

Various changes:
1. Remove unused AddressMap in SPMI. Removed a bunch of cases in
the NearDiffer that were using it.
2. Change arm64 emitter unit tests to not require `verbose`
3. Always build in the late disassembler under DEBUG. It's only
used when `JitLateDisasm` is set.
4. Implement new arm32 callback mechanism for movw/movt handling
in the NearDiffer.

BruceForstall committed:
Update coredistools

Update to the 1.2.0 version of the coredistools package.

Also, implement a basic late disassembler for RyuJIT based on
coredistools (previously, this depended on closed-source msvcdis).

Various changes:
1. Remove unused AddressMap in SPMI. Removed a bunch of cases in
the NearDiffer that were using it.
2. Change arm64 emitter unit tests to not require `verbose`
3. Always build in the late disassembler under DEBUG. It's only
used when `JitLateDisasm` is set.
4. Implement new arm32 callback mechanism for movw/movt handling
in the NearDiffer.

BruceForstall committed:
Allow fallback to coredistools 1.1.0

BruceForstall committed:
Fixes

BruceForstall committed:
Formatting

BruceForstall committed:
Fixes

1. Fix Risc-V build: only build LATE_DISASM for platforms where
coredistool support exists.
2. Defer load and initialization of coredistools.dll library for
LATE_DISASM until we actually need it.
3. For arm64, in late disasm, if there is a failure, try to recover
by skipping one 4-byte instruction.

BruceForstall commented:
/azp run runtime-coreclr outerloop

BruceForstall commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

BruceForstall commented:
gcstress pipeline failure is https://github.com/dotnet/runtime/issues/96364

BruceForstall commented:
@jakobbotsch @dotnet/jit-contrib PTAL

Note this includes changes to work with the 1.3.0 coredistools, but it will also work with the existing 1.1.0 coredistools.

Changing everyone to use 1.3.0 is https://github.com/dotnet/runtime/pull/96291. This still has some problem with linux-x64 gcstress.

The rebuild of cordistools to 1.3.0 is https://github.com/dotnet/jitutils/pull/370. This build still fails in CI with linux-x64 and linux-arm64, but those have been manually built and packaged into an uploaded 1.3.0 coredistools.

kunalspathak commented on src/coreclr/jit/disasm.cpp:
```diff
@@ -1418,12 +1431,12 @@ void DisAssembler::disSetMethod(size_t addr, CORINFO_METHOD_HANDLE methHnd)
     {
         DISASM_DUMP("Helper function: %p => %p\n", addr, methHnd);
-        GetHelperAddrToMethodHandleMap()->Set(addr, methHnd);
+        GetHelperAddrToMethodHandleMap()->Set(addr, methHnd, AddrToMethodHandleMap::SetKind::Overwrite);
```
what are the scenarios with this change that we will have to overwrite the existing entry in a map?

BruceForstall replied:
I think if there are multiple calls to the same function or helper in a method, it will be overwritten (with the same value).

It turns out we currently don't use this in the coredistools-based disassembler.

kunalspathak commented on src/coreclr/jit/disasm.cpp:
```diff
@@ -1544,6 +1562,175 @@ void DisAssembler::disOpenForLateDisAsm(const char* curMethodName, const char* c
+//
+bool DisAssembler::InitCoredistoolsDisasm()
+{
+    // TODO: call g_PtrFinishDisasm sometime?
```
still relevant?

BruceForstall replied:
Yes. I went ahead and wrote some better initialization/un-initialization to fix this. Also fixed it to only load coredistools.dll once per process: it was loading it once for every function being disassembled.

kunalspathak commented on src/coreclr/jit/disasm.cpp:
```diff
@@ -1544,6 +1562,175 @@ void DisAssembler::disOpenForLateDisAsm(const char* curMethodName, const char* c
+
+#ifdef TARGET_UNIX
+    // Unix will require the full path to coredistools. Assume that the
+    // location is next to the full path to the superpmi.so.
```
nit: `./superpmi`. There is no `superpmi.so` on linux.

kunalspathak commented on src/coreclr/jit/disasm.cpp:
```diff
@@ -1544,6 +1562,175 @@ void DisAssembler::disOpenForLateDisAsm(const char* curMethodName, const char* c
+                    codeBytes += 4;
+                    codeSizeBytes -= 4;
+                }
+                if (errorCount < maxErrorCount)
```
shouldn't this check be outside of `#ifdef TARGET_ARM64`?

BruceForstall replied:
no: ARM64 is the only platform currently where there is any logic to recover from a failure to disassemble. For x86/x64, it's not clear what could be done: where would you restart? For arm32, we could skip either 2 or 4 based on an easy check whether the instruction is 2-bytes or 4-bytes, but it didn't seem worth it.

kunalspathak commented on src/coreclr/jit/codegenlinear.cpp:
```diff
@@ -2682,7 +2682,7 @@ void CodeGen::genEmitterUnitTests()
     const WCHAR* unitTestSection = JitConfig.JitDumpEmitUnitTests();
 
-    if ((!verbose) || (unitTestSection == nullptr))
+    if (unitTestSection == nullptr)
```
FYI - @a74nh

BruceForstall replied:
If we want to limit the functions into which we generate the unit tests, I would suggest adding JitEmitUnitTests and make it a `CONFIG_METHODSET` (like JitDisasm/JitDump) to determine which methods will generate the unit tests. (I'd probably rename this to `JitEmitUnitTestsSections`)

Thus, e.g.,
```
DOTNET_JitEmitUnitTests=main
DOTNET_JitEmitUnitTestsSections=sve
```

kunalspathak reviewed (commented):
Just reviewed changes in `jit` folder and they looks good except for few questions/suggestions.

BruceForstall committed:
Be smarter about initializing and uninitializing coredistools

In particular, only load the coredistools library once, not
once per function to disassemble.

BruceForstall committed:
Update to 1.4.0

BruceForstall commented:
superpmi pipeline failures are due to missing osx-arm64 MCH files due to new spmi collection

jakobbotsch commented:
/azp run runtime-coreclr superpmi-replay, runtime-coreclr superpmi-diffs

jakobbotsch commented on src/coreclr/jit/disasm.cpp:
```diff
@@ -37,6 +38,8 @@
 
 /*****************************************************************************/
 
+#ifdef USE_MSVCDIS
```
Do we ever use MSVCDIS? Can we remove all of this?

BruceForstall replied:
We currently don't, but we could resurrect the code and use it. It shouldn't be too hard to do so, and there might be benefits. The only downside is it's not open source. So I'd prefer to just leave it.

jakobbotsch commented on src/coreclr/tools/superpmi/superpmi/neardiffer.cpp:
```diff
@@ -318,147 +341,91 @@ struct DiffData
-    // A new clause that tries to handle the case where neither cr1 or cr2 is the
-    // recorded CR (diff case with two jits, neither of which is the one used for
-    // collection)
+#if 0
+    // We might want to do something similar for mov/movk/movk/movk sequences on Arm64. The following code was
+    // previously used, but currently is not active.
```
Delete the code?

BruceForstall replied:
I'm not convinced yet that we don't need (or will never need) this, so I think it's worth leaving.

jakobbotsch reviewed (approved):
LGTM
